src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,393 @@
+  next_it++;
+  auto end_it = data_to_move_idx.cend();
+  for (; curr_it != end_it; next_it++, curr_it++) {
+    const unsigned int curr_idx = *curr_it;
+    if (next_it == end_it) {
+      if (split) {
+        unique_values[label_old_cluster]->remove_datum(
+            curr_idx, data.row(curr_idx), update_hierarchy_params());
+      } else {
+        remove_singleton(label_old_cluster);
+        /* When a singleton is removed, the other labels are shifted.
+         * Therefore, it is necessary to update label_new_cluster accordingly.
+         */
+        label_new_cluster = allocations[second_random_idx];
+      }
+      unique_values[label_new_cluster]->add_datum(curr_idx, data.row(curr_idx),
+                                                  update_hierarchy_params());
+    } else {
+      unique_values[label_old_cluster]->remove_datum(
+          curr_idx, data.row(curr_idx), false);
+      unique_values[label_new_cluster]->add_datum(curr_idx, data.row(curr_idx),
+                                                  false);
+    }
+    allocations[curr_idx] = label_new_cluster;
+  }
+}
+
+double SplitAndMergeAlgorithm::restricted_gibbs_sampling(
+    const double first_random_idx, bool return_log_res_prod /*=false*/,
+    bool step_to_original_clust /*=false*/) {
+  auto &rng = bayesmix::Rng::Instance().get();
+
+  double log_res_prod = 0;
+
+  for (size_t k = 0; k < restricted_gs_data_idx.size(); ++k) {
+    restricted_gs_unique_values[allocations_restricted_gs[k]]->remove_datum(
+        restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+        update_hierarchy_params());
+
+    Eigen::VectorXd logprobas(2);
+    logprobas(0) = mixing->get_mass_existing_cluster(
+        restricted_gs_data_idx.size() + 2 - 1, true, true,
+        restricted_gs_unique_values[0]);
+    logprobas(0) += restricted_gs_unique_values[0]->conditional_pred_lpdf(
+        data.row(restricted_gs_data_idx[k]));
+    logprobas(1) = mixing->get_mass_existing_cluster(
+        restricted_gs_data_idx.size() + 2 - 1, true, true,
+        restricted_gs_unique_values[1]);
+    logprobas(1) += restricted_gs_unique_values[1]->conditional_pred_lpdf(
+        data.row(restricted_gs_data_idx[k]));
+
+    Eigen::VectorXd probas = stan::math::softmax(logprobas);
+    unsigned int c_new;
+    if (step_to_original_clust) {
+      c_new = (allocations[restricted_gs_data_idx[k]] ==
+               allocations[first_random_idx])
+                  ? 0
+                  : 1;
+    } else {
+      c_new = bayesmix::categorical_rng(probas, rng, 0);
+    }
+
+    if (return_log_res_prod) {
+      log_res_prod += log(probas(c_new));
+    }
+
+    allocations_restricted_gs[k] = c_new;
+    restricted_gs_unique_values[allocations_restricted_gs[k]]->add_datum(
+        restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+        update_hierarchy_params());
+  }
+
+  return log_res_prod;
+}
+
+void SplitAndMergeAlgorithm::full_gibbs_sampling() {
```

Comment:
This code looks very similar to something found in Neal2, but you already discussed this if I recall correctly. If so, please place a "TODO: identical to Neal2" so that we remember to handle this in the future